src/qililab/qprogram/quantum_machines_compiler.py

```diff
@@ -0,0 +1,576 @@
+        if element.variable.domain is Domain.Time:
+            values = np.maximum(values, 4).astype(int)
+        return qua.for_each_(qua_variable, values)
+
+    def _handle_average(self, element: Average):
+        variable = qua.declare(int)
+        return qua.for_(variable, 0, variable < element.shots, variable + 1)
+
+    def _handle_set_frequency(self, element: SetFrequency):
+        bus = self._bus_mapping[element.bus] if self._bus_mapping and element.bus in self._bus_mapping else element.bus
+        frequency = (
+            self._qprogram_to_qua_variables[element.frequency]
+            if isinstance(element.frequency, Variable)
+            else element.frequency * 1e3
+        )
+        qua.update_frequency(element=bus, new_frequency=frequency, units="mHz")
+
+    def _handle_set_phase(self, element: SetPhase):
+        bus = self._bus_mapping[element.bus] if self._bus_mapping and element.bus in self._bus_mapping else element.bus
+        phase = (
+            self._qprogram_to_qua_variables[element.phase]
+            if isinstance(element.phase, Variable)
+            else element.phase / 360.0
+        )
+        qua.frame_rotation_2pi(phase, bus)
+
+    def _handle_reset_phase(self, element: ResetPhase):
+        bus = self._bus_mapping[element.bus] if self._bus_mapping and element.bus in self._bus_mapping else element.bus
+        qua.reset_frame(bus)
+
+    def _handle_set_gain(self, element: SetGain):
+        bus = self._bus_mapping[element.bus] if self._bus_mapping and element.bus in self._bus_mapping else element.bus
+        gain = self._qprogram_to_qua_variables[element.gain] if isinstance(element.gain, Variable) else element.gain
+        # QUA doesn't have a method for setting the gain directly.
+        # Instead, it uses an amplitude multiplication with `amp()`.
+        # Thus, we store the current gain to multiply with in the next play instructions.
+        self._buses[bus].current_gain = gain
+
+    def _handle_play(self, element: Play):
+        bus = self._bus_mapping[element.bus] if self._bus_mapping and element.bus in self._bus_mapping else element.bus
+        waveform_I, waveform_Q = element.get_waveforms()
+        waveform_variables = element.get_waveform_variables()
+        duration = waveform_I.get_duration()
+
+        gain = qua.amp(self._buses[bus].current_gain * 2) if self._buses[bus].current_gain is not None else None
+
+        if not waveform_variables:
+            waveform_I_name = self.__add_waveform_to_configuration(waveform_I)
+            waveform_Q_name = self.__add_waveform_to_configuration(waveform_Q) if waveform_Q else None
+            pulse_name = self.__add_or_update_control_pulse_to_configuration(waveform_I_name, waveform_Q_name, duration)
+            operation_name = self.__add_pulse_to_element_operations(bus, pulse_name)
+            pulse = operation_name * gain if gain is not None else operation_name
+            qua.play(pulse, bus)
+
+    def _handle_measure(
```

Comment:
Yeah, I know this method is huge, but I would like to keep it as it is for now, and not try to break it down/optimize it, mainly because I want all the different branches to be very clear when we try and debug this with the HW people.

Reply:
Fair enough.